### ISSUE TITLE
GlideinWMS fails to import in python 3.6 version of 1.2 branch (because no py3.6 version exists as yet)

### ISSUE THREAD

**Issue by StevenCTimm** [HEPCloud/decisionengine_modules#225]
There are a number of routines in the main glideinwms library that are called by the 
decision engine module glideinwms/transforms/glidein_requests.py 

Historically these have been supplied by the glideinwms-vofrontend-standalone rpm 
and its dependencies.

These python routines are not yet available in python3.  I have filed a ticket with glideinwms support.  I am told they will be available in 3 weeks.

Since there is no python3.6 version of the glideinwms libraries on the machine, the resource_request channel fails to start.

Unit tests didn't catch this.

**Comment by DmitryLitvintsev** [comment]
Marco, besides glideinwms python3 library we need a more meaningful  unit test. It is shame that missing imports have not been caught by existing glideinwms modules tests. 
I am assigning this ticket to you since you are in a good position to write these tests.

**Comment by StevenCTimm** [comment]
According to Marco's presentation in glideinwms stakeholders today he says "End of July" and that Bruno is dedicated to work on this.. the 3.9.0 version will be a python3 version.

Some of the routines we need are evidently among the most difficult to port, namely the ones that do the persistence of the files on the gwms-frontend web server.

**Comment by StevenCTimm** [comment]
I have checked the glideinwms 3.9 branch (which is the python version) out of git and manually installed it into fermicloud117.  It is not released yet but if there are big problems we should be able to find them.

**Comment by StevenCTimm** [comment]
The resource_request channel can now start, all the dependencies are there,  but rapidly goes offline.  Investigating.

**Comment by StevenCTimm** [comment]
note that I had to install pip3 module M2crypto  (which is also on the python 2.7 decision engine) and needed openssl-devel to make that compile.

**Comment by StevenCTimm** [comment]
The resource request channel gets as far as knowing that it needs a particular resource and making classads for it, but the publishers for both glideclientglobals and glideclient fail to advertise to the factory collector.
The collector mapping is correctly set so that fermicloud117 (the test DE) can condor_advertise but it complains that the classad that is being advertised is not well formed.

09/10/20 13:40:43 ZKM: name to map is '/DC=org/DC=incommon/C=US/ST=Illinois/L=Ba
tavia/O=Fermi Research Alliance/OU=Fermilab/CN=fermicloud117.fnal.gov'
09/10/20 13:40:43 ZKM: pre-map: current user is 'gsi'
09/10/20 13:40:43 ZKM: pre-map: current domain is 'unmapped'
09/10/20 13:40:43 ZKM: map file already loaded.
09/10/20 13:40:43 ZKM: attempting to map '/DC=org/DC=incommon/C=US/ST=Illinois/L
=Batavia/O=Fermi Research Alliance/OU=Fermilab/CN=fermicloud117.fnal.gov'
09/10/20 13:40:43 ZKM: 1: attempting to map '/DC=org/DC=incommon/C=US/ST=Illinoi
s/L=Batavia/O=Fermi Research Alliance/OU=Fermilab/CN=fermicloud117.fnal.gov'
09/10/20 13:40:43 ZKM: 2: mapret: 0 included_voms: 0 canonical_user: hepcloudDE1
17
09/10/20 13:40:43 ZKM: found user hepcloudDE117, splitting.
09/10/20 13:40:43 ZKM: post-map: current user is 'hepcloudDE117'
09/10/20 13:40:43 ZKM: post-map: current domain is 'cmssrv258.fnal.gov'
09/10/20 13:40:43 ZKM: post-map: current FQU is 'hepcloudDE117@cmssrv258.fnal.go
v'
09/10/20 13:40:43 AUTHENTICATE: Exchanging keys with remote side.
09/10/20 13:40:43 AUTHENTICATE: Result of end of authenticate is 1.
09/10/20 13:40:43 DC_AUTHENTICATE: authentication of 131.225.155.64 complete.
09/10/20 13:40:43 DC_AUTHENTICATE: message authenticator enabled with key id cms
srv258:2359496:1599763243:140428.
09/10/20 13:40:43 DC_AUTHENTICATE: Success.
09/10/20 13:40:43 IPVERIFY: for fermicloud117.fnal.gov matched 131.225.155.64 to
 131.225.155.64
09/10/20 13:40:43 IPVERIFY: hepcloudDE117@cmssrv258.fnal.gov at <131.225.155.64:
38431> not matched to deny list, so allowing.
09/10/20 13:40:43 PERMISSION GRANTED to hepcloudDE117@cmssrv258.fnal.gov from ho
st 131.225.155.64 for command 58 (UPDATE_AD_GENERIC), access level DAEMON: reaso
n: DAEMON authorization policy does not deny, so allowing; identifiers used for 
this remote host: 131.225.155.64,fermicloud117.fnal.gov
09/10/20 13:40:43 DC_AUTHENTICATE: added incoming session id cmssrv258:2359496:1
599763243:140428 to cache for 80 seconds (lease is 3620s, return address is unkn
own).
09/10/20 13:40:43 Command 58 on Sock not followed by ClassAd (or timeout occured
)
09/10/20 13:40:43 Received malformed ad from command (58). Ignoring.
09/10/20 13:40:45 DaemonCommandProtocol: Not enough bytes are ready for read.
09/10/20 13:40:45 DC_AUTHENTICATE: received DC_AUTHENTICATE from <131.225.155.64
:34942>
09/10/20 13:40:45 DC_AUTHENTICATE: resuming session id cmssrv258:2359496:1599763
243:140428:
09/10/20 13:40:45 SECMAN: other side is $CondorVersion: 8.9.6 Apr 07 2020 BuildI
D: UW_Python_Wheel_Build $, NOT reauthenticating.
09/10/20 13:40:45 DC_AUTHENTICATE: message authenticator enabled with key id cms
srv258:2359496:1599763243:140428.
09/10/20 13:40:45 DC_AUTHENTICATE: Success.
09/10/20 13:40:45 PERMISSION GRANTED to hepcloudDE117@cmssrv258.fnal.gov from ho
st 131.225.155.64 for command 58 (UPDATE_AD_GENERIC), access level DAEMON: reaso
n: cached result for DAEMON; see first case for the full reason
09/10/20 13:40:45 Command 58 on Sock not followed by ClassAd (or timeout occured
)
09/10/20 13:40:45 Received malformed ad from command (58). Ignoring.


I will try to compare the glideclientglobal_manifests data block between python2 and python3 and see what is the difference
This is not easy--both of these data blocks are filled with binary goo.

**Comment by StevenCTimm** [comment]
It is clear that the underlying problem is that the binary hex fields in the glideclientglobal classads
(and probably the regular glideclient classads too) are getting like this "b'13cd1738cc3abc'". i. e. there is a b' '
being encoded around the byte stream values which are just getting put out with a regular print.
The right thing to do according to Marco Mambelli is to encode them with string.encode(binaryvar,"latin-1")
Will need some help from one of the DE developers to make this happen.

**Comment by StevenCTimm** [comment]
To be clear this is not a problem in the new python3 routines that are being imported from the newly-released glideinwms.
It is a change in the way that python3 is behaving and it is part of the codebase that is already part of the DE.  We just never got as far as testing these routines under python3 before.

**Comment by StevenCTimm** [comment]
glideinwms v3_9_0 has now been released.  I have installed the v3_9_0 frontend on cmssrv266 and it works to make classads.  I have also put the v3_9_0 code on fermicloud117 to fulfill the decision engine dependencies and that is (mostly) working too,
to the extent that we have transmitted classads to the test factory cmssrv258 and some glideins (although not all that we thought we asked for) have been submitted.
There is some discrepancy still between the number of idle and total glideins that the DE is sending in its classads, compared to what its logs say it will be requesting.. we are trying to sort this out.
